src/protocol.rs

```diff
@@ -1,42 +1,541 @@
+
+    fn protocol(&self) -> Protocol;
+
+    /// Read address segment data from stream
+    ///
+    /// If the address segment does expect any additional data, then
+    /// no data will be read and an empty struct will be created.
+    fn from_stream(stream: &mut io::Read) -> Result<Self> where Self: Sized;
+
+    /// Generate the canonical binary representation of the contents of this
+    /// address segment
+    ///
+    /// In order to obtain a human-readable string representation use
+    /// `ToString.to_string` instead.
+    fn to_stream(&self, stream: &mut io::Write) -> io::Result<()>;
+}
+
+impl<T> From<T> for Protocol where T: AddressSegment {
+    fn from(addr: T) -> Protocol {
+        addr.protocol()
+    }
+}
+
+
+
+/// A trait for reading any kind of address segment from a byte stream
+pub trait AddressSegmentReaderExt<T: AddressSegment> {
+    fn read_addr(&mut self) -> io::Result<T>;
+}
+impl<T: AddressSegment, R: io::Read> AddressSegmentReaderExt<T> for R {
+    fn read_addr(&mut self) -> io::Result<T> {
+        T::from_stream(self).map_err(|err: Error| {
+            io::Error::new(io::ErrorKind::InvalidData, err)
+        })
+    }
+}
+
+
+/// A trait for writing any kind of address segment to a byte stream
+pub trait AddressSegmentWriterExt<T: AddressSegment> {
+    fn write_addr(&mut self, addr: &T) -> io::Result<()>;
+}
+impl<T: AddressSegment, W: io::Write> AddressSegmentWriterExt<T> for W {
+    fn write_addr(&mut self, addr: &T) -> io::Result<()> {
+        addr.to_stream(self)
+    }
+}
+
+
+macro_rules! derive_for_wrapping_segment {
+    ( $name:ident ) => {};
+    ( $name:ident , $( $rest:tt )+ ) => {
+        derive_for_wrapping_segment!($name : $($rest)*);
+    };
+
+    ( $name:ident : Display $( $rest:tt )* ) => {
+        impl fmt::Display for $name {
+            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+                fmt::Display::fmt(&self.0, f)
+            }
+        }
+
+        derive_for_wrapping_segment!($name $($rest)*);
+    };
+
+    ( $name:ident : FromStr $( $rest:tt )* ) => {
+        impl FromStr for $name {
+            type Err = Error;
+
+            fn from_str(s: &str) -> Result<Self> {
+                Ok($name(FromStr::from_str(s)?))
+            }
+        }
+
+        derive_for_wrapping_segment!($name $($rest)*);
+    };
+
+    ( $name:ident : From < $type:path > $( $rest:tt )* ) => {
+        impl From < $type > for $name {
+            fn from(addr: $type ) -> Self {
+                $name (addr)
+            }
+        }
+
+        derive_for_wrapping_segment!($name $($rest)*);
+    };
+}
+
+
+macro_rules! derive_for_empty_segment {
+    ( $name:ident ) => {};
+    ( $name:ident , $( $rest:tt )+ ) => {
+        derive_for_empty_segment!($name : $($rest)*);
+    };
+
+    ( $name:ident : Display $( $rest:tt )* ) => {
+        impl fmt::Display for $name {
+            fn fmt(&self, _: &mut fmt::Formatter) -> fmt::Result {
+                Ok(())
```

Comment:
I've changed the definition of `AddressSegement` to only require the implementation of `ToString` for each address segement type. Is that good enough?

Otherwise the macro implementation of `Display` for the `Segment` enumeration type will become really painful/impossible:

```rust
        impl fmt::Display for Segment {
            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
                match self {
                    $( &Segment::$var(ref addr) => {
                        f.write_str(addr.protocol().as_str())?;
                        if $addr_type::STREAM_LENGTH > 0 {
                            f.write_char('/')?;
                            f.write_str(addr.to_string().as_ref())?;
                        }

                        Ok(())
                    }, )*

                    // Need extra match arm because of non-exhaustiveness
                    _ => unreachable!()
                }
            }
        }
```